File: src/main/java/com/learning/leetcode/arrays/medium/FindDuplicate.java
package com.learning.leetcode.arrays.medium;

import java.util.Arrays;

public class FindDuplicate {

    public static void main(String[] args) {
        int[] nums = {1,3,4,2,2};
        System.out.println("Given Array: " + Arrays.toString(nums));
        System.out.println("Duplicate Number is " + findDuplicate(nums));
        nums = new int[]{3,1,3,4,2};
        System.out.println("Given Array: " + Arrays.toString(nums));
        System.out.println("Duplicate Number is " + findDuplicate(nums));

    }

    private static int findDuplicate(int[] nums) {
        int slow = 0;
        int fast = 0;
        do {
            slow = nums[slow];
            fast = nums[nums[fast]];
        } while (slow != fast);
        slow = 0;
        while(slow != fast) {
            slow = nums[slow];
            fast = nums[fast];
        }
        return slow;
    }
}

File: src/main/java/com/learning/leetcode/arrays/medium/FindDuplicate.py
from typing import List


def findDuplicate(nums: List[int]) -> int:
    slow = 0
    fast = 0
    while True:
        slow = nums[slow]
        fast = nums[nums[fast]]
        if slow == fast:
            break
    slow = 0
    while slow != fast:
        slow = nums[slow]
        fast = nums[fast]
    return slow


def main():
    nums = [1, 3, 4, 2, 2]
    print("Given Array: ", nums)
    print("Duplicate number ", str(findDuplicate(nums)))
    nums = [3, 1, 3, 4, 2]
    print("Given Array: ", nums)
    print("Duplicate number ", str(findDuplicate(nums)))


main()
